azkeys/coseSigner.go

```diff
@@ -20,32 +20,82 @@ import (
 	"github.com/veraison/go-cose"
 )
 
-// CoseSignerKeyVault is the azure keyvault client for interacting with keyvault keys
-// using a cose.Signer interface
-type CoseSignerKeyVault struct {
-	keyName string
-	alg     keyvault.JSONWebKeySignatureAlgorithm
-	*KeyVault
+// IdentifiableCoseSigner represents a Cose1 signer that has additional methods to provide
+// sufficient information to verify the signed product (an identifier for the signing key and the
+// public key.)
+type IdentifiableCoseSigner interface {
+	cose.Signer
+	PublicKey() (*ecdsa.PublicKey, error)
+	KeyIdentifier() string
+}
+
+// IdentifiableCoseSignerFactory is for creating IdentifiableCoseSigners. The reason for a factory
+// here is that we can always create a fresh instance, capturing the latest key information at that
+// point in time.
+type IdentifiableCoseSignerFactory interface {
+	NewIdentifiableCoseSigner(ctx context.Context) (IdentifiableCoseSigner, error)
+}
+
+// CoseSignerKeyVaultFactory creates instances of our Azure KeyVault implementation of
+// IdentifiableCoseSigner. The keyvault configuration is stored on the object and new instances
+// can be created without caller knowledge of it.
+type CoseSignerKeyVaultFactory struct {
+	keyVaultURL string
+	keyName     string
+}
+
+// NewCoseSignerKeyVaultFactory returns a new instance of the factory, storing the keyvault config
+func NewCoseSignerKeyVaultFactory(keyVaultURL string, keyName string) *CoseSignerKeyVaultFactory {
+	return &CoseSignerKeyVaultFactory{
+		keyVaultURL: keyVaultURL,
+		keyName:     keyName,
+	}
 }
 
 // NewCoseSignerKeyVault creates a new keyvault configuration that signs with ES384
 // using the latest version of the named key
-func NewCoseSignerKeyVault(keyVaultURL string, keyName string) *CoseSignerKeyVault {
+func (f *CoseSignerKeyVaultFactory) NewIdentifiableCoseSigner(ctx context.Context) (IdentifiableCoseSigner, error) {
```

Comment:
Do we not want KeyVault in the name here?  this is not a generic coseSigner, this is a Key Vault one?